lib/internal/buffer.js

```diff
@@ -0,0 +1,87 @@
+'use strict';
+
+if (!process.binding('config').hasIntl) {
+  return;
+}
+
+const normalizeEncoding = require('internal/util').normalizeEncoding;
+const Buffer = require('buffer').Buffer;
+
+const icu = process.binding('icu');
+
+// Maps the supported transcoding conversions. The top key is the fromEncoding,
+// the child key is the toEncoding. The value is the transcoding function.
+const conversions = {
+  'ascii': {
+    'latin1': (source) => {
+      return icu.convert('iso8859-1', 'us-ascii', source);
+    },
+    'utf8': (source) => {
+      return icu.convert('utf-8', 'us-ascii', source);
+    },
+    'utf16le': (source) => {
+      return icu.convertToUcs2('us-ascii', source);
+    }
+  },
+  'latin1': {
+    'ascii': (source) => {
+      return icu.convert('us-ascii', 'iso8859-1', source);
+    },
+    'utf8': (source) => {
+      return icu.convert('utf-8', 'iso8859-1', source);
+    },
+    'utf16le': (source) => {
+      return icu.convertToUcs2('iso8859-1', source);
+    }
+  },
+  'utf8': {
+    'ascii': (source) => {
+      return icu.convert('us-ascii', 'utf-8', source);
+    },
+    'latin1': (source) => {
+      return icu.convert('iso-8859-1', 'utf-8', source);
+    },
+    'utf16le': icu.convertToUcs2FromUtf8
+  },
+  'utf16le': {
+    'ascii': (source) => {
+      if (source.length % 2 !== 0)
+        throw new TypeError('Invalid UCS2 Buffer');
+      return icu.convertFromUcs2('us-ascii', source);
+    },
+    'latin1': (source) => {
+      if (source.length % 2 !== 0)
+        throw new TypeError('Invalid UCS2 Buffer');
+      return icu.convertFromUcs2('iso-8859-1', source);
+    },
+    'utf8': (source) => {
+      if (source.length % 2 !== 0)
+        throw new TypeError('Invalid UCS2 Buffer');
+      return icu.convertToUtf8FromUcs2(source);
+    }
+  }
+};
+
+// Transcodes the Buffer from one encoding to another, returning a new
+// Buffer instance.
+exports.transcode = function transcode(source, fromEncoding, toEncoding) {
+  if (!source || !(source.buffer instanceof ArrayBuffer))
+    throw new TypeError('"source" argument must be a Buffer');
+  if (source.length === 0) return Buffer.alloc(0);
+
+  fromEncoding = normalizeEncoding(fromEncoding) || fromEncoding;
+  toEncoding = normalizeEncoding(toEncoding) || toEncoding;
+
+  if (fromEncoding === toEncoding)
+    return Buffer.from(source);
```

Comment:
The transcoder is set up to replace the invalid character sequences when going from one encoding to another. I'm not particularly concerned with that in regards to this short cut.